src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,263 @@
+ *
+ * @expiremental in 5.1
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+final class AutoMapper implements AutoMapperInterface, AutoMapperRegistryInterface, MapperGeneratorMetadataRegistryInterface
+{
+    /** @var MapperGeneratorMetadataInterface[] */
+    private $metadata = [];
+
+    /** @var GeneratedMapper[] */
+    private $mapperRegistry = [];
+
+    private $classLoader;
+
+    private $mapperConfigurationFactory;
+
+    public function __construct(ClassLoaderInterface $classLoader, MapperGeneratorMetadataFactoryInterface $mapperConfigurationFactory = null)
+    {
+        $this->classLoader = $classLoader;
+        $this->mapperConfigurationFactory = $mapperConfigurationFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register(MapperGeneratorMetadataInterface $metadata): void
+    {
+        $this->metadata[$metadata->getSource()][$metadata->getTarget()] = $metadata;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMapper(string $source, string $target): MapperInterface
+    {
+        $metadata = $this->getMetadata($source, $target);
+
+        if (null === $metadata) {
+            throw new NoMappingFoundException('No mapping found for source '.$source.' and target '.$target);
+        }
+
+        $className = $metadata->getMapperClassName();
+
+        if (\array_key_exists($className, $this->mapperRegistry)) {
+            return $this->mapperRegistry[$className];
+        }
+
+        if (!class_exists($className)) {
+            $this->classLoader->loadClass($metadata);
+        }
+
+        $this->mapperRegistry[$className] = new $className();
+        $this->mapperRegistry[$className]->injectMappers($this);
+
+        foreach ($metadata->getCallbacks() as $property => $callback) {
+            $this->mapperRegistry[$className]->addCallback($property, $callback);
+        }
+
+        return $this->mapperRegistry[$className];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasMapper(string $source, string $target): bool
+    {
+        return null !== $this->getMetadata($source, $target);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map($sourceData, $targetData, array $context = [])
+    {
+        $source = null;
+        $target = null;
+
+        if (null === $sourceData) {
+            return null;
+        }
+
+        if (\is_object($sourceData)) {
+            $source = \get_class($sourceData);
+        }
+
+        if (\is_array($sourceData)) {
+            $source = 'array';
+        }
+
+        if (null === $source) {
+            throw new NoMappingFoundException('Cannot map this value, source is neither an object or an array.');
+        }
+
+        if (\is_object($targetData)) {
+            $target = \get_class($targetData);
+            $context[MapperContext::TARGET_TO_POPULATE] = $targetData;
+        }
+
+        if (\is_array($targetData)) {
```

Comment:
Why not `elseif` here, `$targetData` cannot be an object and an array at the same time.

Reply:
Both works and would have the same effect on peformance, more an habit of code here don't know the rule of Symfony here ?

Reply:
>Both works and would have the same effect on peformance

It would not, using `elseif`s would prevent evaluating the following conditions as soon as one is true, preventing function calls for type checks in this case.

Reply:
Hum that's true, feel stupid xD, will change this, thanks.